src/smartWalletDiscovery.ts

```diff
@@ -0,0 +1,321 @@
+  const config = {
+    ...DEFAULT_DISCOVERY_CONFIG,
+    ...configOveride,
+  };
+
+  const wordlist = wordlists[config.mnemonicLanguage];
+  const hardenedNode: HDNode = HDNode.fromMnemonic(
+    mnemonic,
+    password,
+    wordlist
+  );
+  const provider = getDefaultProvider(config.network);
+
+  const smartWalletFactory = SmartWalletFactory__factory.connect(
+    config.factory,
+    provider
+  );
+
+  const creationByteCode = await smartWalletFactory.getCreationBytecode();
+  const bytecodeHash = utils.keccak256(creationByteCode);
+  const logic = config.logic ?? constants.AddressZero;
+  const logicParamsHash = config.logicParamsHash ?? SHA3_NULL_S;
+
+  return {
+    config: {
+      ...config,
+      logic,
+      logicParamsHash,
+    },
+    hardenedNode,
+    bytecodeHash,
+    smartWalletFactory,
+    provider,
+  };
+}
+
+async function findActiveSWFor(
+  eoaAddr: string,
+  config: Config,
+  bytecodeHash: string,
+  provider: providers.Provider
+) {
+  let inactivityCountdown = config.sWalletGap;
+  let swIndex = 0;
+  let swAccounts: string[] = [];
+
+  while (inactivityCountdown >= 0) {
+    // originally <=gapLimit, but rskj rpc crashes
+    const currentSWAddress = getSWAddress(
+      config,
+      eoaAddr,
+      swIndex,
+      bytecodeHash
+    );
+
+    const hasActivity = await hasAccActivity(
+      currentSWAddress,
+      config,
+      provider
+    );
+
+    if (hasActivity) {
+      swAccounts = [...swAccounts, currentSWAddress];
+      inactivityCountdown = config.sWalletGap;
+    } else {
+      inactivityCountdown -= 1;
+    }
+
+    swIndex++;
+  }
+
+  return swAccounts;
+}
+
+const discoverAccountsFromMnemonic = async (
+  configOveride: Config,
+  mnemonic: string,
+  password?: string
+): Promise<Account[]> => {
+  const { config, hardenedNode, bytecodeHash, provider } = await setupDiscovery(
+    configOveride,
+    mnemonic,
+    password
+  );
+
+  let accounts: Account[] = [];
+  let hardenedAccount = -1;
+  let nextHardened = true;
+
+  while (nextHardened) {
+    hardenedAccount += 1;
+    nextHardened = false;
+    let inactivityCountdown = config.eoaGap;
+    let eoaIndex = 0;
+
+    while (inactivityCountdown > 0) {
+      const { address: eoaAccount } = hardenedNode.derivePath(
+        constructDerivationPath(eoaIndex, hardenedAccount)
+      );
+      const eoaActivityFound = await hasAccActivity(
+        eoaAccount,
+        config,
+        provider
+      );
+
+      if (eoaActivityFound) {
+        nextHardened = true;
+        inactivityCountdown = config.eoaGap;
+      }
+
+      const swAccounts = await findActiveSWFor(
+        eoaAccount,
+        config,
+        bytecodeHash,
+        provider
+      );
+      
+      if (swAccounts.length > 0 || eoaActivityFound) {
+        nextHardened = true;
+        inactivityCountdown = config.eoaGap;
+
+        accounts = [
+          ...accounts,
+          {
+            eoaAccount,
+            swAccounts,
+          },
+        ];
```

Comment:
It's not clear to me if we're covering [this case](https://github.com/rsksmart/rif-relay-client/blob/master/src/SmartWalletDiscovery.ts#L223)

Reply:
yes, that's only there becuase they initialise the `discoveredAccount` object and then mutate it, whereas we use it directly.

Reply:
You're right, my fault. Thank you for the clarification 👍🏼